clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,485 @@
+		return
+	}
+	klog.Infof("Stopping network performance measurement...")
+	if err := npm.deleteCRD(); err != nil {
+		klog.Errorf("error while deleting CRD: %v", err)
+	}
+
+	if err := npm.deleteClusterRoleBinding(); err != nil {
+		klog.Errorf("error while deleting rbac: %v", err)
+	}
+	if err := client.DeleteNamespace(npm.k8sClient, netperfNamespace); err != nil {
+		klog.Errorf("error while deleting %s namespace: %v", netperfNamespace, err)
+	}
+	if err := client.WaitForDeleteNamespace(npm.k8sClient, netperfNamespace); err != nil {
+		klog.Errorf("error while waiting for %s namespace to be deleted: %v", netperfNamespace, err)
+	}
+
+	npm.closeCh()
+}
+
+func (npm *networkPerformanceMeasurement) watchNetworkTestEvents() {
+	dynamicClient := npm.framework.GetDynamicClients().GetClient()
+	gvk := schema.GroupVersionKind{Group: customResourceGroup, Kind: customResourceKind, Version: customResourceVersion}
+	gvr, _ := meta.UnsafeGuessKindToResource(gvk)
+	watchResource, err := dynamicClient.Resource(gvr).Watch(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		klog.Error("Error watching", err)
+		return
+	}
+	eventChannel := watchResource.ResultChan()
+	for event := range eventChannel {
+		var metricResp metricResponse
+		if event.Type != watch.Modified {
+			continue
+		}
+		unstructuredList, _ := event.Object.(runtime.Unstructured)
+		unstructuredObjContent := unstructuredList.UnstructuredContent()
+		metricMap := unstructuredObjContent["status"].(map[string]interface{})
+		getMetricResponse(metricMap, &metricResp)
+		if len(metricResp.Metrics) == 0 && metricResp.Error != "" {
+			klog.Error("error message from worker pod: ", metricResp.Error)
+			continue
+		}
+		npm.metricLock.Lock()
+		npm.metricVal = append(npm.metricVal, metricResp.Metrics)
+		npm.metricLock.Unlock()
+	}
+}
+
+func (npm *networkPerformanceMeasurement) execNToMTest() {
+	go npm.createCustomResourcePerUniquePodPair()
+
+	npm.formUniquePodPair()
+}
+
+func (npm *networkPerformanceMeasurement) createCustomResourcePerUniquePodPair() {
+	for CRIndex := 0; ; CRIndex++ {
+		uniqPodPair := <-npm.podPairCh
+		if uniqPodPair.isLastPodPair {
+			break
+		}
+		if npm.startTimeStampForTestExecution == 0 {
+			npm.startTimeStampForTestExecution = getStartTimeStampForPod()
+		}
+		npm.templateMapping = npm.populateTemplate(uniqPodPair, CRIndex)
+		if err := npm.framework.ApplyTemplatedManifests(networkTestRequestFilePath, npm.templateMapping); err != nil {
+			klog.Error(err)
+		}
+		npm.podPairNumber++
+	}
+}
+
+func (npm *networkPerformanceMeasurement) formUniquePodPair() {
+	var uniqPodPair uniquePodPair
+	lastPodPair := uniquePodPair{isLastPodPair: true}
+
+	var unUsedPodIndex = 0
+	for {
+		for workerNode, podList := range npm.workerPodList {
+			unUsedPod, err := getUnusedPod(&podList)
+			npm.workerPodList[workerNode] = podList
+			if err != nil {
+				delete(npm.workerPodList, workerNode)
+				continue
+			}
+			unUsedPodIndex++
```

Comment:
implementing 2 phases by for seems strange

It would be much cleaner to have:

for ; ; len(l) > 0 {
  first, err := l.Pop()
  // handle err
  second, err := l.Pop()
  // handle err

  podPair := uniquePodPair{
    sourcePodIP: first.podIP
    ...
  }
```

Reply:
Our requirement is create a unique pair of pods residing on two different worker nodes. Thats why i have created a map which stores list of pods per worker node name. So the logic which is implemented above is:
a) Choose one unused pod from 1st worker node and update the pod list in that worker node by reducing it by one.
b) Then move it to next worker node and choose the pod from that list. Again  update the pod list in that worker node by reducing it by one.

This loop goes on until all pods are exhausted.

Reply:
Still - this logic should be hidden in the struct that is wrapping that map.
Sth like:

for ; ; len(l) > 0 {
first, err := l.Pop()
// handle err
second, err := l.PopFromADifferentNodeThan(first))
// handle err

podPair := uniquePodPair{
sourcePodIP: first.podIP